src/kernel/src/dmem/mod.rs

```diff
@@ -1,40 +1,138 @@
+use thiserror::Error;
+
+use crate::dev::{Dmem, DmemContainer};
 use crate::errno::EINVAL;
-use crate::fs::Fs;
-use crate::process::VThread;
+use crate::fs::{make_dev, CharacterDevice, DriverFlags, Fs, MakeDevError, MakeDevFlags, Mode};
+use crate::info;
+use crate::process::{VProc, VThread};
 use crate::syscalls::{SysErr, SysIn, SysOut, Syscalls};
+use crate::ucred::{Gid, Privilege, Uid};
+use std::ops::Index;
 use std::sync::Arc;
 
 pub use self::blockpool::*;
 
 mod blockpool;
 
+#[derive(Debug)]
+pub struct DmemDevice {
+    name: &'static str,
+    dev: Arc<CharacterDevice>,
+}
+
+impl DmemDevice {
+    pub const NAMES: [&'static str; 3] = ["dmem0", "dmem1", "dmem2"];
+
+    pub(super) fn new(name: &'static str, dev: Arc<CharacterDevice>) -> Self {
+        Self { name, dev }
+    }
+}
+
 /// An implementation of direct memory system on the PS4.
 pub struct DmemManager {
     fs: Arc<Fs>,
+    dmem0: DmemDevice,
+    dmem1: DmemDevice,
+    dmem2: DmemDevice,
 }
 
 impl DmemManager {
+    const DMEM_TOTAL_SIZE: usize = 0x13C_000_000;
+
     pub fn new(fs: &Arc<Fs>, sys: &mut Syscalls) -> Arc<Self> {
-        let dmem = Arc::new(Self { fs: fs.clone() });
+        let dmem0 = {
+            let name = "dmem0";
+            match make_dev(
+                Dmem::new(Self::DMEM_TOTAL_SIZE, DmemContainer::Zero),
+                DriverFlags::D_INIT,
+                0,
+                name,
+                Uid::ROOT,
+                Gid::ROOT,
+                Mode::new(0o777).unwrap(),
+                None,
+                MakeDevFlags::empty(),
+            ) {
+                Ok(v) => Ok(DmemDevice::new(name, v)),
+                Err(e) => Err(DmemManagerInitError::CreateDmemFailed(name, e)),
+            }
+        }
+        .expect("error creating dmem0");
```

Comment:
Create a DmemInitError here or something like that and use ?; It will also require to create the corresponding variant of KernelError in main.rs.